3 improviements to transforms: 1. helper function to get an affine tr…

LucaMarconato opened the pull request:
…ansform from input and output coordinate system; 2. fixed but in sequence (now inferring coordinate systems from its components); 3. added missing test for inferencen of intermediate coordinate systems for nested sequences

LucaMarconato commented:
As mentioned in my previous post, a downside of " setter for elements to check that the transfomration makes sense (if not, tries to fix it)" is that very verbose transformations are generated.

For instance this code here:

LucaMarconato commented:
After discussing with @giovp I will refactor into separate classes to divide what is purely a NGFF transform from what is built on top of it for ergonomics. The reasons are the following:
1) the NGFF classes can be in principle moved to `ome-zarr-py` or an object-oriented representation of the storage, which is something that we want;
2) this approach will greatly simplify the code complexity. The NGFF classes will be used only when reading or writing. The in-memory representation will take advantage of the new classes.
3) The new classes will be based on `xarray` like described here https://github.com/scverse/spatialdata/issues/47 and the order of axes will be irrelevant. Also, all the transformations will have non-ambiguous information for the axis (on the contrary NGFF Sequence transform allow for containing transformations which don't specify the axes. This increases the code complexity.
4) New functionalities that will be easily supported will be the possibility to concatenate any sequence of transformations independently of the axes they operate on, and assign any transformation to any element, again independently of the axes they work on.

The old version of the code can be found in the [previous commit](https://github.com/scverse/spatialdata/pull/100/commits/20f2e3a79bb3417db7c6dadaa5e93b68237f43b4), as well as in a [separate branch](https://github.com/scverse/spatialdata/tree/archive/only_ngff_transforms) created from it. If the new code works as planned I'll delete the branch (which I already tagged for archiving purposes and [can be found here](https://github.com/scverse/spatialdata/commit/20f2e3a79bb3417db7c6dadaa5e93b68237f43b4)).

LucaMarconato commented:
Also, before the review I would suggest to look at this other PR and merge it to main, so that the diff is a bit smaller: https://github.com/scverse/spatialdata/pull/113

giovp commented on spatialdata/_core/_spatialdata.py:
```diff
@@ -242,12 +280,391 @@ def _init_add_element(self, name: str, element_type: str, overwrite: bool) -> za
+        self._write_transformations_to_disk(element)
+
+    @staticmethod
+    def remove_transformation_in_memory(
```
same for this and below

giovp commented on spatialdata/_core/_spatialdata.py:
```diff
@@ -242,12 +280,391 @@ def _init_add_element(self, name: str, element_type: str, overwrite: bool) -> za
+        self.remove_transformation_in_memory(element, target_coordinate_system)
+        self._write_transformations_to_disk(element)
+
+    @staticmethod
```
same again for this, I think there should be 1 method only. Also, when do we need to "remove" transformations>?

giovp commented on spatialdata/_core/_spatialdata.py:
```diff
@@ -242,12 +280,391 @@ def _init_add_element(self, name: str, element_type: str, overwrite: bool) -> za
+                src_node = id(source_coordinate_system)
+            else:
+                src_node = source_coordinate_system
+            des_node: Union[int, str]
```
`des_node` -> `tgt_node` ? although I like the german influence 😅

LucaMarconato replied:
changed, both you and copilot voted for `tgt_node` ahah

giovp commented on spatialdata/_core/_spatialdata.py:
```diff
@@ -242,12 +280,391 @@ def _init_add_element(self, name: str, element_type: str, overwrite: bool) -> za
+                        path = paths_with_length_1[0]
+                    else:
+                        # error 1
+                        raise RuntimeError(
+                            "Multiple paths found between the two coordinate systems. Please specify an intermediate "
+                            "coordinate system."
+                        )
```
maybe it's worth here to print the path to tie it to what went wrong?

LucaMarconato replied:
Good point, added it. Now the exception looks like this:
```
>                       raise RuntimeError(
                            "Multiple paths found between the two coordinate systems. Please specify an intermediate "
                            f"coordinate system. Available paths are:{s}"
                        )
E                       RuntimeError: Multiple paths found between the two coordinate systems. Please specify an intermediate coordinate system. Available paths are:
E                           my_space0 -> /images/image2d_multiscale -> global
E                           my_space0 -> /labels/labels2d -> global
```

LucaMarconato replied:
Actually I changed it to this so it's clearer what the user should plug in the function call:
```
E                       RuntimeError: Multiple paths found between the two coordinate systems. Please specify an intermediate coordinate system. Available paths are:
E                           my_space0 -> <sdata>.images['image2d_multiscale'] -> global
E                           my_space0 -> <sdata>.labels['labels2d'] -> global
```

giovp commented on spatialdata/_core/_spatialdata.py:
```diff
@@ -242,12 +280,391 @@ def _init_add_element(self, name: str, element_type: str, overwrite: bool) -> za
+                elements[element_type][element_name] = element
+        return SpatialData(**elements, table=self.table)
+
+    def transform_element_to_coordinate_system(
```
I would remove this as API wise I don't super get it, will elaborate in comment

LucaMarconato replied:
Ok thanks, answering in the comment.

LucaMarconato replied:
I have removed these functions and added new global methods with better signature (I put them in `_spatialdata.py`, but probably we want to move them somewhere when we will refactor the import structure of the subpackages.

giovp reviewed (changes requested):
So one major comment I have so far (by just checking the `_spatialdata.py` file) is the type of API we want to have. Copying it from the design doc

```python
import spatialdata as sd
from spatialdata import SpatialData

sdata = SpatialData(...)
points = sd.transform(sdata.points["image1"], tgt="tgt_space")
sdata = sd.transform(sdata, tgt="tgt_space")
```

I think we should have 1 way to apply transform to either spatialdata or elements, and maybe we also shouldnt' allow user to pass directly the element, but only the name and type (e.g. {"labels":"sample1"}. I think this would prevent mistakes in copy/viewes of elements from the ones already in spatialdata and the ones transformed. wdyt ?

LucaMarconato commented on spatialdata/_core/_spatialdata.py:
```diff
@@ -242,12 +280,391 @@ def _init_add_element(self, name: str, element_type: str, overwrite: bool) -> za
+        transformations[target_coordinate_system] = transformation
+        _set_transformations(element, transformations)
+
+    def set_transformation(
+        self,
+        element: SpatialElement,
+        transformation: BaseTransformation,
+        target_coordinate_system: Optional[str] = None,
+    ) -> None:
+        """
+        Set/replace a transformation of an element, writing it to disk if the SpatialData object is backed.
+
+        Parameters
+        ----------
+        element
+            The element to replace the transformation for.
+        transformation
```
Unfortunately not, one is a static method, the other is a normal method.

LucaMarconato replied:
In particular, the non-static method takes a `SpatialElement` as input and finds it inside the `SpatialData` object. In this way we can support changing also the disk content automatically. The disk is updated when `sdata` is backed, that is when `sdata.path` is not `None`.

LucaMarconato replied:
replace with better methods

LucaMarconato commented:
Here is the example notebook showcasing transformations (and also a bit of io, mostly related to transformations).

https://github.com/scverse/spatialdata-notebooks/blob/main/notebooks/transformations_examples.ipynb

LucaMarconato commented on pyproject.toml:
```diff
@@ -70,6 +71,8 @@ xfail_strict = true
 addopts = [
 #    "-Werror",  # if 3rd party libs raise DeprecationWarnings, just use filterwarnings below
     "--import-mode=importlib",  # allow using test files with same name
+    "-x",  # exit on first test if it fails
```
Right, removed.

LucaMarconato commented on spatialdata/_core/_spatialdata.py:
```diff
@@ -179,28 +200,45 @@ def _add_labels_in_memory(self, name: str, labels: Union[SpatialImage, Multiscal
+    def _get_group_for_element(self, name: str, element_type: str) -> zarr.Group:
+        store = parse_url(self.path, mode="r+").store
+        root = zarr.group(store=store)
+        assert element_type in ["images", "labels", "points", "polygons", "shapes"]
```
added a todo in the code

LucaMarconato commented on spatialdata/_core/_spatialdata.py:
```diff
@@ -242,12 +280,391 @@ def _init_add_element(self, name: str, element_type: str, overwrite: bool) -> za
+                raise ValueError("Unknown element type")
+
+    @staticmethod
+    def get_all_transformations(element: SpatialElement) -> MappingToCoordinateSystem_t:
```
Maybe we could do it. I would keep the default behaviour of target_coordinate_system to be `None` (well actually "global", gonna change it now). We can have an argument `return_all: bool = False` and change the return type to `Union[BaseTransformation, dict[str, BaseTransformation]`.

LucaMarconato replied:
We could also make `target_coordinate_system: Optional[str] = DEFAULT_COORDINATE_SYSTEM` (="global") and return all the transformations if it is set to None by the user (but it feels a bit weird).

LucaMarconato replied:
I have removed these functions and added new global methods with better signature (I put them in `_spatialdata.py`, but probably we want to move them somewhere when we will refactor the import structure of the subpackages.

LucaMarconato commented:
> So one major comment I have so far (by just checking the _spatialdata.py file) is the type of API we want to have. Copying it from the design doc
> 
> import spatialdata as sd
> from spatialdata import SpatialData
> 
> sdata = SpatialData(...)
> points = sd.transform(sdata.points["image1"], tgt="tgt_space")
> sdata = sd.transform(sdata, tgt="tgt_space")
> I think we should have 1 way to apply transform to either spatialdata or elements, and maybe we also shouldnt' allow user to pass directly the element, but only the name and type (e.g. {"labels":"sample1"}. I think this would prevent mistakes in copy/viewes of elements from the ones already in spatialdata and the ones transformed. wdyt ?

I think that we need two functions:
1. one for getting a transformation between one target and one source, and this would need to be a method of `SpatialData` since it needs to build the digraph of the transformations of the elements that are inside. This is currently the function `map_coordinate_systems()`. We can change the name, it's a bit confusing. Maybe `map()`?
2. one method to actually transform one element to a coordinate system. Again this needs to be a method of `SpatialData` because it needs to call the function above internally.

This is the reason why I couldn't go with this idea from the design doc:
```
points = sd.transform(sdata.points["image1"], tgt="tgt_space")
```

Still we could change the second function to this:
```
points = sd.transform(sdata, element_type='labels', element_name='cells', tgt="tgt_space")
```

Here are the pro and cons of this second approach.
Pro:
* we can rewrite and `set_transformation()`, `remove_transformation()` and `remove_all_transformations()` ato take as first argument the `sdata` object and then the string description of the element like above. In this way we can remove the `*_in_memory()` methods since all the functions can be static methods (or just global functions). I really like this.
Con:
* passing both the `element_type` and the `element_name` is (ihmo very) cumbersome. One way to get around this would be to require unique names for elements. I explore this possibility in this issue that I have just created: https://github.com/scverse/spatialdata/issues/124

My proposal is to require unique names for elements and after that to change the second function as described. I would do this in a separate pr (I don't think we will have problems of copy/views, things should work right now).

kevinyamauchi reviewed (approved):
Thanks @LucaMarconato ! I think this is a solid step in the right direction. I am approving so that we can keep things moving forward. I think we will have to continue to iterate on API, etc., but it will be easier to do so when we actually use it to make vignettes, etc.

> one for getting a transformation between one target and one source, and this would need to be a method of SpatialData since it needs to build the digraph of the transformations of the elements that are inside. This is currently the function map_coordinate_systems(). We can change the name, it's a bit confusing. Maybe map()?

Perhaps it could be called something like `get_transformation_between_coordinate_systems()`? Probably too long, but more descriptive. In any case, I agree that `map_coordinate_systems()` is probably not the most clear name. 

Does this method actually need to be on `SpatialData`? It seems to me that one could still construct the directed graph of the transformations on a given element. Perhaps I'm missing something though.

LucaMarconato commented:
Thanks for the comments @kevinyamauchi, I have removed the functions `get`/`set`/`remove` transformations (total of 8 functions) and replaced with 3 functions living outside the `SpatialData` class. I have also moved `map_coordinate_systems` out of the `SpatialData` class and renamed it to `get_transformation_between_coordinate_systems()`.

LucaMarconato commented:
I will now merge the new points io pr, it will take some times to fix all the tests, but then we'll be ready to merge!

giovp commented:
@LucaMarconato quick thing I noticed from working in #132 : the `PointsModel.parse` still requires an `annotation` to be passed in the case of `nd.array` input. I think this should not be enforced since the `feature_key` is not mandatory. Was it fixed here or shall I push a quick fix in main?

LucaMarconato commented:
Hi, true I didn't fix it actually. A quick fix would be great thanks.
